xapian-core/weight/weight.cc

```diff
@@ -134,4 +135,81 @@ Weight::unserialise(const string &) const
     throw Xapian::UnimplementedError("unserialise() not supported for this Xapian::Weight subclass");
 }
 
+const Weight *
+Weight::create(const string & s, const Registry & reg)
+{
+    const char *p = s.c_str();
+    std::string scheme = "";
+
+    while (*p != ' ') {
+	if (*p == '\0') break;
+	scheme += *p;
+	p++;
+    }
+
+    if (scheme == "bb2")
+	scheme = "Xapian::BB2Weight";
+    else if (scheme == "bm25")
+	scheme = "Xapian::BM25Weight";
+    else if (scheme == "bm25plus")
+	scheme = "Xapian::BM25PlusWeight";
+    else if (scheme == "bool")
+	scheme = "Xapian::BoolWeight";
+    else if (scheme == "coord")
+	scheme = "Xapian::CoordWeight";
+    else if (scheme == "dlh")
+	scheme = "Xapian::DLHWeight";
+    else if (scheme == "dph")
+	scheme = "Xapian::DPHWeight";
+    else if (scheme == "ifb2")
+	scheme = "Xapian::IfB2Weight";
+    else if (scheme == "ineb2")
+	scheme = "Xapian::IneB2Weight";
+    else if (scheme == "inl2")
+	scheme = "Xapian::InL2Weight";
+    else if (scheme == "pl2")
+	scheme = "Xapian::PL2Weight";
+    else if (scheme == "pl2plus")
+	scheme = "Xapian::PL2PlusWeight";
+    else if (scheme == "tfidf")
+	scheme = "Xapian::TfIdfWeight";
+    else if (scheme == "trad")
+	scheme = "Xapian::TradWeight";
```

Comment:
It's not good to hard-code this list - the idea is this works for user schemes too.  Also it's `O(n)` lookup.

I'd overlooked that the current key here is the class name.  I think we need a `map` (or perhaps better `unordered_map`) to track the schemes by the short name too.